server/src/main/java/org/opensearch/extensions/action/ExtensionActions.java

```diff
@@ -0,0 +1,149 @@
+        this.client = client;
+    }
+
+    void registerAction(String action, DiscoveryExtension extension) throws IllegalArgumentException {
+        if (actionsMap.containsKey(action)) {
+            throw new IllegalArgumentException("The " + action + " you are trying to register is already registered");
+        }
+        actionsMap.putIfAbsent(action, extension);
+    }
+
+    public DiscoveryExtension getExtension(String action) {
+        return actionsMap.get(action);
+    }
+
+    /**
+     * Handles a {@link RegisterTransportActionsRequest}.
+     *
+     * @param transportActionsRequest  The request to handle.
+     * @return  A {@link ExtensionBooleanResponse} indicating success.
+     */
+    public TransportResponse handleRegisterTransportActionsRequest(RegisterTransportActionsRequest transportActionsRequest) {
+        /*
+         * We are proxying the transport Actions through ExtensionMainAction, so we really dont need to register dynamic actions for now.
+         */
+        logger.debug("Register Transport Actions request recieved {}", transportActionsRequest);
+        DiscoveryExtension extension = extensionIdMap.get(transportActionsRequest.getUniqueId());
+        for (String action : transportActionsRequest.getTransportActions().keySet()) {
+            registerAction(action, extension);
+        }
+        return new ExtensionBooleanResponse(true);
+    }
+
+    public TransportResponse handleTransportActionRequestFromExtension(TransportActionRequestFromExtension request) {
+        DiscoveryExtension extension = extensionIdMap.get(request.getUniqueId());
+        final CountDownLatch inProgressLatch = new CountDownLatch(1);
+        final TransportActionResponseToExtension response = new TransportActionResponseToExtension(new byte[0]);
+        client.execute(
+            ExtensionMainAction.INSTANCE,
+            new ExtensionActionRequest(request.getAction(), request.getRequestBytes()),
+            new ActionListener<ExtensionActionResponse>() {
+                @Override
+                public void onResponse(ExtensionActionResponse actionResponse) {
+                    response.setResponseBytes(actionResponse.getResponseBytes());
+                    inProgressLatch.countDown();
+                }
+
+                @Override
+                public void onFailure(Exception exp) {
+                    logger.debug("Transport request failed", exp);
+                    byte[] responseBytes = ("Request failed: " + exp.getMessage()).getBytes(StandardCharsets.UTF_8);
+                    response.setResponseBytes(responseBytes);
+                    inProgressLatch.countDown();
+                }
+            }
+        );
+        return response;
+    }
+
+    public ExtensionActionResponse sendTransportRequestToExtension(ExtensionActionRequest request) {
+        DiscoveryExtension extension = actionsMap.get(request.getAction());
+        final CountDownLatch inProgressLatch = new CountDownLatch(1);
```

Comment:
Similar comment here: where are we awaiting this countdown?

Reply:
Same as above, we really would like to wait until the response comes back before we respond.